### PR TITLE
Simplify Readers and Writers

### PR DESCRIPTION
**Pull request by zeroZshadow**
Simplify codegen for readers and writers by using shared abstract base classes

### CODE REVIEW

File: workers/unity/Packages/io.improbable.gdk.core/Subscriptions/ReaderSubscriptionManager.cs [hunk 1]
```diff
@@ -5,7 +5,7 @@
 namespace Improbable.Gdk.Subscriptions
 {
     public abstract class ReaderSubscriptionManager<TComponent, TReader> : SubscriptionManager<TReader>
-        where TReader : IRequireable
+        where TReader : IRequireable where TComponent : ISpatialComponentData
```
Comment (paulbalaji) [inline workers/unity/Packages/io.improbable.gdk.core/Subscriptions/ReaderSubscriptionManager.cs]:
```suggestion
        where TReader : IRequireable
        where TComponent : ISpatialComponentData
```

Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.core/Subscriptions/ReaderSubscriptionManager.cs]: please split these + others like these across lines

Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.core/Subscriptions/ReaderSubscriptionManager.cs]: general style question: should it be ordered TComponent before TReader?

File: workers/unity/Packages/io.improbable.gdk.core/Subscriptions/Reader.cs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        where TComponent : struct, IComponentData, ISpatialComponentData
+        where TUpdate : struct, ISpatialComponentUpdate
+    {
+        public bool IsValid { get; set; }
+        protected readonly ComponentUpdateSystem ComponentUpdateSystem;
+        protected readonly ComponentCallbackSystem CallbackSystem;
+        protected readonly EntityManager EntityManager;
+        protected readonly Entity Entity;
+        protected readonly EntityId EntityId;
+        private readonly uint componentId = ComponentDatabase.GetComponentId<TComponent>();
```
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.core/Subscriptions/Reader.cs]: Nit: could be static

File: workers/unity/Packages/io.improbable.gdk.core/Subscriptions/ReaderSubscriptionManager.cs (hunk 1 above)
Comment (jamiebrynes7) [inline workers/unity/Packages/io.improbable.gdk.core/Subscriptions/ReaderSubscriptionManager.cs]: For completeness, should `TComponent : struct` as well?

**Comment by zeroZshadow** [comment]
Made all componentID's that were added static, so it will only be looked up once per type

File: workers/unity/Packages/io.improbable.gdk.core/Subscriptions/WriterSubscriptionManager.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
+using System.Collections.Generic;
+using Improbable.Gdk.Core;
+using Improbable.Worker.CInterop;
+using Unity.Entities;
+using Entity = Unity.Entities.Entity;
+
+namespace Improbable.Gdk.Subscriptions
+{
+    public abstract class WriterSubscriptionManager<TComponent, TWriter> : SubscriptionManager<TWriter>
+        where TWriter : IRequireable where TComponent : ISpatialComponentData
```
Comment (paulbalaji) [inline workers/unity/Packages/io.improbable.gdk.core/Subscriptions/WriterSubscriptionManager.cs]:
```suggestion
        where TWriter : IRequireable
        where TComponent : ISpatialComponentData
```

**Pull request merged**